crates/iota-indexer/src/apis/governance_api.rs

```diff
@@ -337,12 +345,98 @@ impl<T: R2D2Connection + 'static> GovernanceReadApi<T> {
+
+        let mut cache = self.exchange_rates_cache.lock().await;
+
+        // Check if the exchange rates for the current epoch are cached
+        if let Some(cached_rates) = cache.cache_get(&epoch) {
+            return Ok(cached_rates.clone());
+        }
+
+        // Cache miss: compute exchange rates
+        let exchange_rates = self.compute_exchange_rates(system_state_summary).await?;
+
+        // Store in cache
+        cache.cache_set(epoch, exchange_rates.clone());
+
+        Ok(exchange_rates)
+    }
+
+    /// Compute Exchange Rates for Active & Inactive validators
+    async fn compute_exchange_rates(
+        &self,
+        system_state_summary: &IotaSystemStateSummary,
+    ) -> Result<Vec<ValidatorExchangeRates>, IndexerError> {
+        // Get validator rate tables
+        let mut tables = vec![];
+
+        for validator in &system_state_summary.active_validators {
+            tables.push((
+                validator.iota_address,
+                validator.staking_pool_id,
+                validator.exchange_rates_id,
+                validator.exchange_rates_size,
+                true,
+            ));
+        }
+
+        // Get inactive validator rate tables
+        for df in self
+            .inner
+            .get_dynamic_fields_in_blocking_task(
+                system_state_summary.inactive_pools_id,
+                None,
+                system_state_summary.inactive_pools_size as usize,
+            )
+            .await?
+        {
+            let pool_id: iota_types::id::ID = bcs::from_bytes(&df.bcs_name).map_err(|e| {
+                iota_types::error::IotaError::ObjectDeserialization {
+                    error: e.to_string(),
+                }
+            })?;
+            let inactive_pools_id = system_state_summary.inactive_pools_id;
+            let validator = self
+                .inner
+                .spawn_blocking(move |this| {
+                    iota_types::iota_system_state::get_validator_from_table(
+                        &this,
+                        inactive_pools_id,
+                        &pool_id,
+                    )
+                })
+                .await?;
+            tables.push((
+                validator.iota_address,
+                validator.staking_pool_id,
+                validator.exchange_rates_id,
+                validator.exchange_rates_size,
+                false,
+            ));
+        }
+
+        self.validator_exchange_rates(tables).await
+    }
+
+    /// Check if there is any `Pending` and `Candidate` validators and get its
+    /// exchange rates, this two states of a validator lifecycle can manifest
+    /// during an epoch or multiple ones, is essential to not cache this data,
+    /// while exchange rates for `Active` and `Inactive` validators is OK to do
+    /// so because those states are manifested on epoch change.
+    pub async fn pending_and_candidate_validators_exchange_rate(
```

Comment:
Again I would suggest splitting the method between `pending` and `candidate` to minimize the scope.

Reply:
As suggested the [two functions](https://github.com/iotaledger/iota/pull/4313/commits/353da3b8f1c64730650f552d6b69a8a80833f0d9#diff-324f865071b593cb4ed59dd865b9600fae68065a818e622b14f4f5554812fd98R177-R187) now are separate